core/domain/expression_parser.py

```diff
@@ -0,0 +1,509 @@
+class Token(python_utils.OBJECT):
+    """Class for tokens of the math expression."""
+
+    def __init__(self, text):
+        """Initializes a Token object.
+
+        Args:
+            text: str. String representation of the token.
+        """
+        self.text = text
+
+        # Categorize the token.
+        if self.is_identifier(text):
+            self.category = 'identifier'
+        elif self.is_function(text):
+            self.category = 'function'
+        elif self.is_number(text):
+            self.category = 'number'
+        elif self.is_operator(text):
+            self.category = 'operator'
+        else:
+            raise Exception('Invalid syntax')
+
+
+    def is_function(self, text):
+        """Checks if given token represents a valid math function.
+
+        Args:
+            text: str. String representation of the token.
+        Returns:
+            bool. Whether the given string represents a valid math function.
+        """
+        return text in MATH_FUNCTIONS
+
+
+    def is_identifier(self, text):
+        """Checks if given token represents a valid identifier. A valid
+        identifier could be a single latin letter (uppercase/lowercase), a greek
+        letter represented by the symbol name, or a valid math function.
+
+        Args:
+            text: str. String representation of the token.
+        Returns:
+            bool. Whether the given string represents a valid identifier.
+        """
+        if text.isalpha() and len(text) == 1:
+            return True
+        if text in GREEK_LETTERS:
+            return True
+        return False
+
+    def is_number(self, text):
+        """Checks if given token represents a valid number(integer/float)
+        without a '+'/'-' sign.
+
+        Args:
+            text: str. String representation of the token.
+        Returns:
+            bool. Whether the given string represents a valid number.
+        """
+        return text.replace('.', '', 1).isdigit()
+
+    def is_operator(self, text):
+        """Checks if given token represents a valid operator.
+
+        Args:
+            text: str. String representation of the token.
+        Returns:
+            bool. Whether the given string represents a valid operator.
+        """
+        return text in VALID_OPERATORS
+
+
+class Node(python_utils.OBJECT):
+    """Instances of this class act as nodes of the parse tree."""
+
+    def __init__(self, operator_token, children):
+        """Initializes a Node object. For ex. 'a + b' will have root node as
+        '+' (operator_token) and children as ['a', 'b'].
+
+        Args:
+            operator_token: Token. The token representing the operator.
+            children: list(Node). Child nodes of the current node.
+        """
+        self.operator_token = operator_token
+        self.children = children
+
+
+class AddOperator(Node):
+    """Class representing the addition operator node."""
+
+    def __init__(self, operator_token, left, right):
```

Comment:
Yes, it would be the same for every instance of this class. But, let's say in the future we need to perform some operations on this parse tree which requires iterating over the nodes of the tree. We would then need some information about which operator/identifier the node contains, right? We can't just have the node be represented by its children only. There has to be some information about the value that node stores.

Reply:
Sure, but you don't need to pass it in right? You could still set it as e.g. a class constant.

Reply:
I understand. Done.